core/src/test/java/com/graphhopper/routing/ch/PrepareContractionHierarchiesTest.java

```diff
@@ -473,7 +473,7 @@ public void testStallOnDemandViaVirtuaNode_issue1574() {
         // would not happen, because node 2 would not even be explored in the forward search, but because of the virtual
         // node the strict upward search is modified and goes like 0-3-1-2 (i.e. it finds node 2).
         // so no we fine tune the weight for 2-4 such that node 4 gets also stalled
-        edge24.setReverse(carEncoder.getAverageSpeedEnc(), 27.5);
+        edge24.set(carEncoder.getAverageSpeedEnc(), 27.5);
```

Comment:
@easbar this was necessary to make the test pass. 

When I allow both directions via:

```
new CarFlagEncoder(new PMap("speed_two_directions=true"))
```

then I get 4 shortcuts (not 2) and a different path:

```
java.lang.AssertionError: wrong or no path found
Expected :[0, 3, 8, 1, 2, 4, 5, 6, 7]
Actual   :[0, 3,    1, 2, 4, 5, 6, 7]
```

Reply:
Ok this is strange. setting `speed_two_directions` only allows setting two different speeds right ? all the edges are bidirectional anyway. And what I really do not understand is that the calculated path seems to skip the (virtual) node 8 (probably because some shortcut is used, but this should not be). I will take a closer look to see what is going on here.

Reply:
Ok I took a look at this test and fixed some inaccurate comments, see here: ec01e5e458bce46ffddea8a7b266cacbc49e6a1e.
I also explicitly set the speed for *both* directions in this test now, see here: 
c979cbd4511a3263bc6ec312c4a55ba6a1ec1c20

> then I get 4 shortcuts (not 2) 

Ok this is clear, because here we are only counting the number of shortcut edges (independent of their access flags). so far `set()` set the speed for both directions, but with the new code only one of the directions is set, right ? if the edge speeds are different for the two directions the previous two 'bidirectional' shortcuts will turn into four unidirectional ones.

If I comment out the assertEquals(there are two shortcuts), the next thing that fails is this:
`sptWeight at node 4 should be smaller than shortcut weight 3->4 to make sure node 4 gets stalled`
This is because you changed `setReverse` to `set` and the `getWeight` method read the wrong weight (it encounters the 'wrong' edge first and returns it), I fixed this here: a5297537fb5bd15bd9a7c512bc494d503ab9a670

>  And what I really do not understand is that the calculated path seems to skip the (virtual) node 8 

This is because when we use `setReverse` instead of `set` the edge weights are not such that the stalling happens. When the best path is found via the 3->4 shortcut the virtual node will not be included in the final path (because when the shortcut is unpacked it does not know about `QueryGraph`s virtual node). 

`setReverse` was wrong in the test so far, but so far this was no problem, because it did the same as `set` (it set *both* directions).

The best way to check everything is fine is reverting the #1574 fix (set precision back to 0 in the stalling condition in `DijkstraBidirectionCH`) and make sure no path is found. Also run `RandomCHRouting#random` many times as described in it's comment, but be aware that it currently does not use `speed_both_directions=true`.

Reply:
If you update to master no changes should be required for the test to work and setting `speed_both_directions=true` should be fine.